### PR TITLE
Support for lego's timeout option

### PR DESCRIPTION
**Pull request by tehscorpion**
I hope i've done nothing wrong but at least it compiles and runs :8ball: 
#1034

### CODE REVIEW

File: caddytls/client.go [hunk 1]
```diff
@@ -60,6 +61,11 @@ var newACMEClient = func(config *Config, allowPrompts bool) (*ACMEClient, error)
 		return nil, fmt.Errorf("%s: insecure CA URL (HTTPS required)", caURL)
 	}
 
+	// Set timeout by replacing ACME's http.Client
+	if HTTPTimeout > 1 {
+		acme.HTTPClient = http.Client{Timeout: time.Duration(HTTPTimeout) * time.Second}
```
Comment (mholt) [inline caddytls/client.go]: This should probably be assigned only once, not ever time we create a new ACME client. @xenolf can you confirm when you have a chance?

File: caddy/caddymain/run.go [hunk 2]
```diff
@@ -34,6 +34,7 @@ func init() {
 	flag.StringVar(&cpu, "cpu", "100%", "CPU cap")
 	flag.BoolVar(&plugins, "plugins", false, "List installed plugins")
 	flag.StringVar(&caddytls.DefaultEmail, "email", "", "Default ACME CA account email address")
+	flag.IntVar(&caddytls.HTTPTimeout, "catimeout", 10, "ACME CA HTTP timeout in seconds")
```
Comment (mholt) [inline caddy/caddymain/run.go]: This should be a DurationVar. Then you don't have to convert it when we actually use it.

File: caddytls/tls.go [hunk 3]
```diff
@@ -157,6 +157,9 @@ var (
 	// DefaultEmail represents the Let's Encrypt account email to use if none provided.
 	DefaultEmail string
 
+	// Sets the maximum waiting time for the CA's HTTP responses
+	HTTPTimeout int
```
Comment (mholt) [inline caddytls/tls.go]: As above, change to time.Duration.

File: caddytls/client.go (hunk 1 above)
Comment (xenolf) [inline caddytls/client.go]: It's a package level variable and not connected to a single client instance. Assigning it once is enough.

Comment (mholt) [inline caddytls/client.go]: Okay, thanks. @tehscorpion can you make this so that it's only assigned once?

**Comment by tehscorpion** [comment]
I changed it according to your request, but i have actually no idea where to put it...at least where you want it.
@mholt could you refine it according to your preference?

**Comment by mholt** [comment]
Thanks @tehscorpion! I'll comment inline.

File: caddytls/client.go [hunk 4]
```diff
@@ -60,6 +61,9 @@ var newACMEClient = func(config *Config, allowPrompts bool) (*ACMEClient, error)
 		return nil, fmt.Errorf("%s: insecure CA URL (HTTPS required)", caURL)
 	}
 
+	// Set timeout by replacing ACME's http.Client
+	acme.HTTPClient = http.Client{Timeout: HTTPTimeout}
```
Comment (mholt) [inline caddytls/client.go]: This creates a new http client every time we perform an ACME request. Actually, what if we just changed the timeout value? `acme.HTTPClient.Timeout = HTTPTimeout` - then we don't have to replace the whole client each time.

Comment (mholt) [inline caddytls/client.go]: Better yet, just remove this line and change the flag line to be:

```
flag.DurationVar(&acme.HTTPClient.Timeout, "catimeout", acme.HTTPClient.Timeout, "Default ACME CA HTTP timeout")
```

**Comment by mholt** [comment]
Thank you!

**Pull request merged**